### ISSUE TITLE
PKI not setting 'extendedKeyUsage = serverAuth' when 'server_flag=true'

### ISSUE THREAD

**Issue by ericchapman** [hashicorp/vault#847]
Hi Vault Team,

I am using Vault's PKI to sign certificates for my system and noticed that my generated certificates do not allow me to access Consul BUT I know my certificate is valid because it allows me to access Vault.  Here is my setup.

  - I followed this tutorial to create a CA and intermediate CA to sign my own certificates https://jamielinux.com/docs/openssl-certificate-authority/index.html
  - I generated a certificate for Consul using the intermediate CA with the following settings
    - extendedKeyUsage = clientAuth, serverAuth (per Consul instructions)
    - Common Name = server.\<data center\>.\<domain\>
  - I setup my Consul cluster with TLS enabled and the systems cluster no problem
  - I generate a user certificate to access Consul using "extendedKeyUsage = clientAuth"
  - I generate a P12 for that
  - Using Curl, I verify that I CANNOT connect to Consul on port 8501 (HTTPS) using the P12
  - I generate a new user certificate to access Consul using "extendedKeyUsage = clientAuth, serverAuth"
  - I generate a P12 for that
  - Using Curl, I verify that I CAN connect to Consul on port 8501 (HTTPS) using the P12
  - I standup the Vault cluster using TLS and same certs as Consul (per Vault instructions)
  - I initialize and unseal the Vault
  - I verify Vault is working using curl and my good P12 by calling the health check.  All good
  - I verify that the system does not work if I do not include the certificate meaning TLS is definitely enabled

So at this point, my system is fully bootstrapped and I have verified that everything is working correctly using TLS and my good certificate.  Now to use PKI

I configure vault as follows

    # Mount the pki
    curl -k --cert $1 -H "X-Vault-Token:${VAULT_TOKEN}" \
        -X POST -d "{\"type\": \"pki\"}" \
        $leader/v1/sys/mounts/pki

    # Configure the CA
    curl -k --cert $1 -H "X-Vault-Token:${VAULT_TOKEN}" \
        -X POST --data "@cabundle.json" \
        $leader/v1/pki/config/ca

    role=$(echo ${DOMAIN} | sed 's/\./-/g')
    echo "Creating role: ${role}"

    # Create the roles
    curl -k --cert $1 -H "X-Vault-Token:${VAULT_TOKEN}" \
    -X POST -d "{ \"allowed_base_domain\": \"${DOMAIN}\", \"allow_subdomains\":true, \"max_ttl\":\"72h\", \"server_flag\": true, \"client_flag\": true }" \
    $leader/v1/pki/roles/$role

I then issue a certificate as follows

    curl -k --cert $1 -H "X-Vault-Token:${VAULT_TOKEN}" \
        -X POST -d "{\"common_name\": \"${3}.${DOMAIN}\"}" \
        $leader/v1/pki/issue/$role

I get back the private key, CA, and certificate.  I bundle those up into a P12.  I then use it to access the Vault.

    curl -k --cert $1 https://${SERVER_BOOTSTRAP_IP}:8200/v1/sys/health

This works.  I then use it to access Consul

    curl -k --cert $1 https://${SERVER_BOOTSTRAP_IP}:8501/v1/status/leader

This does not work.  This is behaving identically to if I generate my initial user certificate with "extendedKeyUsage = clientAuth" and no "serverAuth".

My hypothesis is that the generated certificate is not generated with "serverAuth" given that it is behaving the same as the certificate that I explicitly generated without server auth (works with Vault, does not work with Consul).  2 questions

  - How do I confirm this?
  - If this is the case, what more can I do other than set "server_flag=true" during the configuration of the PKI?  I also noticed this was the default but for debug I decided to make it explicit.

Thanks

**Comment by jefferai** [comment]
Hi there,

Try using `openssl x509 -in <cert> -noout -text`. Using this, I see the following from a cert issued using mostly default role parameters (including the defaults for server auth and client auth):
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            30:9f:d6:e2:64:45:7f:ba:0e:3f:8d:34:5a:74:61:65:ca:4b:66:e9
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=intermediate.com
        Validity
            Not Before: Dec 13 21:24:03 2015 GMT
            Not After : Jan 12 21:24:03 2016 GMT
        Subject: CN=localhost
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (224 bit)
                pub: 
                    04:97:a4:66:49:e6:73:94:7b:68:ee:52:3c:e2:80:
                    29:7c:76:81:08:20:1b:b7:7f:3f:63:41:20:4d:a8:
                    4a:61:a2:b1:e6:ea:36:18:5c:e1:a7:74:cf:f0:c5:
                    ae:63:3b:a5:58:cd:c5:ce:ac:22:3b:ff
                ASN1 OID: secp224r1
                NIST CURVE: P-224
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Key Agreement
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Subject Key Identifier: 
                16:3D:81:BA:AB:53:C0:F9:1C:01:3D:3E:A2:75:C3:58:AA:CE:C9:54
            X509v3 Authority Key Identifier: 
                keyid:FF:22:22:39:4B:84:89:7C:22:39:D3:81:A5:A7:FB:68:7E:4F:D8:7B

            X509v3 Subject Alternative Name: 
                DNS:localhost, DNS:foo.example.com, DNS:bar.example.com, IP Address:128.3.5.6, IP Address:FE01:0:0:0:0:0:0:1
    Signature Algorithm: sha256WithRSAEncryption
         a2:58:ec:47:4f:e0:c8:2f:dd:6e:4e:56:34:d3:1f:9f:06:1d:
         2a:73:78:ab:95:7d:cc:f9:fd:ef:1f:3b:52:92:ab:6a:1d:33:
         81:ca:6f:96:12:8e:ee:17:64:d9:96:67:15:15:90:14:52:7e:
         30:cd:5f:9c:0c:82:84:f6:ae:6c:22:95:33:48:c4:a9:f0:06:
         4f:4a:5e:cf:61:09:51:87:ca:1e:bb:70:14:1b:cb:38:55:4d:
         96:c3:b2:2f:b7:c6:0c:44:9b:fa:39:c8:c4:bb:ae:65:8f:2d:
         f5:83:be:59:86:20:f5:5b:39:92:8b:8e:28:3b:b7:87:c5:83:
         92:50:62:72:48:1f:59:89:25:15:d9:4c:fe:cf:81:d3:99:8b:
         02:3d:ff:a6:dc:b3:6e:9e:32:ef:f3:6d:59:54:a0:2d:e1:b3:
         ac:10:e6:9b:41:ee:f5:d6:eb:70:cc:f0:54:b4:75:38:53:bc:
         92:83:04:0c:96:1c:09:62:0e:24:00:fb:cd:18:47:12:f3:17:
         50:55:1e:d1:e3:f6:fb:86:fd:a4:60:89:d9:c9:1c:22:7f:b3:
         63:7d:91:88:ed:2c:cf:73:60:03:b2:2e:ff:79:c6:5c:21:1d:
         90:96:f1:10:6a:2c:88:93:a7:54:27:91:c4:5b:c7:fd:28:6b:
         2e:de:72:a3
```
You can see that web and server auth are both in the extended key usages. What are you seeing?

**Comment by jefferai** [comment]
I'm closing this as it could not be reproduced and may be fixed by other fixes to the PKI backend in the 0.5 dev cycle. Please request it to be reopened if necessary.
